homeassistant/components/zwave_js/api.py

```diff
@@ -786,3 +796,129 @@ async def get(self, request: web.Request, config_entry_id: str) -> web.Response:
                 hdrs.CONTENT_DISPOSITION: 'attachment; filename="zwave_js_dump.json"',
             },
         )
+
+
+@websocket_api.require_admin  # type: ignore
+@websocket_api.async_response
+@websocket_api.websocket_command(
+    {
+        vol.Required(TYPE): "zwave_js/abort_firmware_update",
+        vol.Required(ENTRY_ID): str,
+        vol.Required(NODE_ID): int,
+    }
+)
+@async_get_node
+async def websocket_abort_firmware_update(
+    hass: HomeAssistant,
+    connection: ActiveConnection,
+    msg: dict,
+    node: Node,
+) -> None:
+    """Abort a firmware update."""
+    await node.async_abort_firmware_update()
+    connection.send_result(msg[ID])
+
+
+@websocket_api.require_admin  # type: ignore
+@websocket_api.async_response
+@websocket_api.websocket_command(
+    {
+        vol.Required(TYPE): "zwave_js/subscribe_firmware_update_status",
+        vol.Required(ENTRY_ID): str,
+        vol.Required(NODE_ID): int,
+    }
+)
+@async_get_node
+async def websocket_subscribe_firmware_update_status(
+    hass: HomeAssistant,
+    connection: ActiveConnection,
+    msg: dict,
+    node: Node,
+) -> None:
+    """Subsribe to the status of a firmware update."""
+
+    @callback
+    def async_cleanup() -> None:
+        """Remove signal listeners."""
+        for unsub in unsubs:
+            unsub()
+
+    @callback
+    def forward_progress(event: dict) -> None:
+        progress: FirmwareUpdateProgress = event["firmware_update_progress"]
+        connection.send_message(
+            websocket_api.event_message(
+                msg[ID],
+                {
+                    "sent_fragments": progress.sent_fragments,
+                    "total_fragments": progress.total_fragments,
+                },
+            )
+        )
+
+    @callback
+    def forward_finished(event: dict) -> None:
+        finished: FirmwareUpdateFinished = event["firmware_update_finished"]
+        connection.send_message(
+            websocket_api.event_message(
+                msg[ID],
+                {
+                    "status": finished.status,
+                    "wait_time": finished.wait_time,
+                },
+            )
+        )
+
+    unsubs = [
+        node.on("firmware update progress", forward_progress),
+        node.on("firmware update finished", forward_finished),
+    ]
+    connection.subscriptions[msg["id"]] = async_cleanup
+
+    connection.send_result(msg[ID])
+
+
+class FirmwareUploadView(HomeAssistantView):
+    """View to upload firmware."""
+
+    url = "/api/zwave_js/firmware/upload/{config_entry_id}/{node_id}"
```

Comment:
You can add regex to limit what matches in the url.
```suggestion
    url = "/api/zwave_js/firmware/upload/{config_entry_id}/{node_id:\d+}"
```

That way you don't have to handle parsing errors for `node_id`.